src/librustc_mir/borrow_check/nll/region_infer/mod.rs

```diff
@@ -147,6 +148,40 @@ pub struct Constraint {
 
 newtype_index!(ConstraintIndex { DEBUG_FORMAT = "ConstraintIndex({})" });
 
+/// Constraints that are considered interesting can be categorized to
+/// determine why they are interesting.
+#[derive(Debug, Eq, PartialEq)]
+enum ConstraintCategory {
+    Assignment,
+    CallArgument,
+    Cast,
+    Other,
+}
+
+impl Ord for ConstraintCategory {
+    fn cmp(&self, other: &Self) -> Ordering {
+        if self == other {
+            return Ordering::Equal;
+        }
+
+        match (self, other) {
```

Comment:
Can we not just `#[derive(Ord)]` here? I think that would work fine...?

Reply:
(We would want to comment that the order is significant)

Reply:
I considered doing that, but it would have meant that variants at the top of the definition were considered most significant - which seemed a little too implicit? - so I went with this instead. I'll change it to `#[derive(Ord)]`.